scale_build/package.py

```diff
@@ -0,0 +1,167 @@
+
+    deps_mapping = {
+        p.name: {d for d in p.build_time_dependencies() if d not in built} for p in list(to_build.values())
+    }
+    sorted_ordering = [list(deps) for deps in toposort(deps_mapping)]
+
+    for item in filter(
+        lambda i: i in to_build and i not in in_progress and i not in package_queue.queue and i not in built,
+        sorted(sorted_ordering[0], key=lambda k: to_build_orig[k].batch_priority) if sorted_ordering else []
+    ):
+        package_queue.put(to_build_orig.pop(item))
+
+
+def build_package(package_queue, to_build, failed, in_progress, built):
+    while True:
+        if not failed and (to_build or package_queue.queue):
+            try:
+                package = package_queue.get(timeout=5)
+            except queue.Empty:
+                package = None
+            else:
+                in_progress[package.name] = package
+        else:
+            if PKG_DEBUG:
+                logger.debug('Thread exiting')
+            break
+
+        if package:
+            try:
+                logger.debug('Building %r package', package.name)
+                package.delete_overlayfs()
+                package.setup_chroot_basedir()
+                package.make_overlayfs()
+                with APT_LOCK:
+                    package.clean_previous_packages()
+                    shutil.copytree(PKG_DIR, package.dpkg_overlay_packages_path)
+                package._build_impl()
+            except Exception as e:
+                logger.error('Failed to build %r package', package.name)
+                failed[package.name] = {'package': package, 'exception': e}
+                break
+            else:
+                with APT_LOCK:
+                    package.logger.debug('Building local APT repo Packages.gz...')
+                    run(
+                        f'cd {PKG_DIR} && dpkg-scanpackages . /dev/null | gzip -9c > Packages.gz',
+                        shell=True, logger=package.logger
+                    )
+                in_progress.pop(package.name)
+                built[package.name] = package
+                logger.info(
+                    'Successfully built %r package (Remaining %d packages)', package.name,
+                    len(to_build) + package_queue.qsize() + len(in_progress)
+                )
+
+        with PACKAGE_BUILD_LOCK:
+            if not package:
+                update_queue(package_queue, to_build, failed, in_progress, built)
+
+
+def build_packages():
+    clean_bootstrap_logs()
+    _build_packages_impl()
+
+
+def _build_packages_impl():
+    package_logger = get_logger('build_packages', 'build_packages.log', 'w')
+    logger.info('Building packages (%s/build_packages.log)', LOG_DIR)
+    logger.debug('Setting up bootstrap directory')
+
+    with PackageBootstrapDirectory(package_logger) as p:
+        p.setup()
+
+    logger.debug('Successfully setup bootstrap directory')
+
+    shutil.rmtree(PKG_LOG_DIR, ignore_errors=True)
```

Comment:
What errors are you ignoring?

We expect `PKG_LOG_DIR` to be non-existent after this method terminates successfully. Will it always be so?